cryostatio/cryostat3
# feat(custom_targets): handle embedded credentials, enable tests

# Welcome to Cryostat3! 👋
## Before contributing, make sure you have:

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Fixes #101
Related to #122
Based on #183
Depends on #183

## Description of the change:
Re-enables the Custom Targets integration test that was ported over earlier. Along with this come some fixes to the initial 3.0 implementation so that the API responses match the pre-existing format, addition of handling of "embedded credentials" supplied with Custom Targets definitions, addition of handling to look up existing credentials from the keyring when opening a target connection, fixes to ensure that WebSocket messages are emitted in a more predictable order for clients to observe, some basic refactoring, and putting back the three instances of `vertx-fib-demo` like the old smoketest has.

## How to manually test:
1. Build PR locally, `./mvnw clean verify ; podman image prune -f` and ensure all tests pass
2. `./smoketest.bash -OXgtb` or whichever combination of flags you like
3. Go to Dashboard > Create Target view, try to create a new custom target. This will need to have a unique but existing/reachable JMX URL, so you can use `service:jmx:rmi:///jndi/rmi://vertx-fib-demo-2:9094/jmxrmi`. In the latest smoketest script, `vertx-fib-demo-n` and `sample-app-n` (`n: [1, 2, 3]`) are basically synonymous - both forms are resolvable hostnames from the perspective of the cryostat3 container. `vertx-fib-demo-2:9094` and `vertx-fib-demo-3:9095` are free to use as custom targets, the other URLs will be automatically discovered by JDP and the samples' Agent instances.
![image](https://github.com/cryostatio/cryostat3/assets/3787464/60c3c1b7-14df-4f3b-b582-80e5547f67ea)

With `service:jmx:rmi:///jndi/rmi://vertx-fib-demo-2:9094/jmxrmi`:
1. Go to target creation view, enter this URL
2. Test connection. It should fail due to missing credentials.
3. Add the credentials `admin:adminpass123` in the form, test the connection again. It should succeed.
4. Replace the credentials with `user:password`, test again. It should still succeed.
5. Replace the credentials with something else, test again. It should fail.
6. Put back `user:password` and create the custom target. This should succeed and some notifications should appear. You will be brought back to the Topology view, which depends on GraphQL (#11), so don't worry about that view being broken for now.
7. Go to Recordings and try to create a new flight recording. This should fail with an HTTP 403.
8. Go to Security and delete the credentials that match the JMX URL. Create new credentials with the same match expression, but this time using the `admin:adminpass123` combination.
9. Go back to Recordings and retry creating a recording. This should succeed.

With `service:jmx:rmi:///jndi/rmi://vertx-fib-demo-3:9095/jmxrmi`:
1. Go to target creation view, enter this URL
2. Test connection. It should fail due to "non-JRMP endpoint" - this is a little misleading but actually indicates the unknown SSL cert required by this test app instance.
3. Creating the target should similarly fail.

With `service:jmx:rmi:///jndi/rmi://cryostat3:9091/jmxrmi`:
1. Go to target creation view, enter this URL
2. Test connection. This should succeed.
3. Creating the target should succeed. Cryostat should succeed at automatically connecting to itself and via this URL you should also see the `onstart` recording that is already present.

Note: *deleting* a Custom Target is only exposed in the UI via the Topology view, so there is no way to undo this graphically yet. You can either tear down and restart the smoketest to get a fresh slate, or you can use curl/HTTPie and delete the custom target directly via the API:

```bash
$ http --auth=user:pass :8181/api/v3/targets
HTTP/1.1 200 OK
Cache-Control: no-cache
Content-Type: application/json;charset=UTF-8
content-encoding: gzip
content-length: 1762

[
    ...
    {
        "agent": false,
        "alias": "service%3Ajmx%3Armi%3A%2F%2F%2Fjndi%2Frmi%3A%2F%2Fvertx-fib-demo-2%3A9094%2Fjmxrmi",
        "annotations": {
            "cryostat": {
                "REALM": "Custom Targets"
            },
            "platform": {}
        },
        "connectUrl": "service:jmx:rmi:///jndi/rmi://vertx-fib-demo-2:9094/jmxrmi",
        "id": 9,
        "jvmId": "hk58o_x3gJT4jK0aLmd5XdBRrdIRKmHb3xKZKixAQDI=",
        "labels": {}
    }
]
```

```bash
$ http --auth=user:pass DELETE :8181/api/v3/targets/9
HTTP/1.1 204 No Content
content-encoding: identity
```
